### PR TITLE
Add `chainerx.accuracy`

### PR DESCRIPTION
**Pull request by aksub99**
Add `chainerx.accuracy`

### CODE REVIEW

File: chainerx_cc/chainerx/python/routines.cc [hunk 1]
```diff
@@ -983,7 +995,7 @@ void InitChainerxPooling(pybind11::module& m) {
 
 void InitChainerxRoutines(pybind11::module& m) {
     InitChainerxCreation(m);
-    InitChainerxIndexing(m);
+    InitChainerxEvaluation(m) : InitChainerxIndexing(m);
```
Comment (take-cheeze) [inline chainerx_cc/chainerx/python/routines.cc]: Typo of colon and semi-colon?

Comment (aksub99) [reply chainerx_cc/chainerx/python/routines.cc]: Thanks @take-cheeze . Fixed it.

**Comment by aksub99** [comment]
The forward pass seems to be working properly now, but the backward still fails.

**Comment by aksub99** [comment]
The implementation seems to be working fine now. I get the same error as in #7524 though.
`AttributeError: 'NoneType' object has no attribute 'shape'`
It seems as though the gradients are nonzero even when the outputs are zero.
@hvy Could you please review?

**Comment by aksub99** [comment]
@hvy Were you able to take a look? The gradients seem to be nonzero even though the outputs are `None`. I don't understand why such a case is arising.

**Comment by hvy** [comment]
You're testing backward for accuracy. This routine is not differentiable so you can skip the backwards tests using `skip_backward_test` and `skip_double_backward_test`.

File: chainerx_cc/chainerx/routines/evaluation.cc [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+#include "chainerx/routines/logic.h"
+#include "chainerx/routines/manipulation.h"
+#include "chainerx/routines/misc.h"
+#include "chainerx/routines/reduction.h"
+#include "chainerx/routines/sorting.h"
+#include "chainerx/routines/statistics.h"
+#include "chainerx/routines/type_util.h"
+#include "chainerx/scalar.h"
+#include "chainerx/shape.h"
+#include "chainerx/strides.h"
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/evaluation.cc]: Can you make sure only necessary headers are included?

File: chainerx_cc/chainerx/routines/evaluation.h [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <cstdint>
+#include <string>
+
+#include <absl/types/optional.h>
+
+#include "chainerx/array.h"
+#include "chainerx/constant.h"
```
[the remark this answers is not in the document]
Comment (aksub99) [reply chainerx_cc/chainerx/routines/evaluation.h]: Fixed.

File: chainerx_cc/chainerx/routines/evaluation.h [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <cstdint>
+#include <string>
```
[the remark this answers is not in the document]
Comment (aksub99) [reply chainerx_cc/chainerx/routines/evaluation.h]: Fixed.

File: chainerx_cc/chainerx/routines/evaluation.cc (hunk 2 above)
Comment (aksub99) [reply chainerx_cc/chainerx/routines/evaluation.cc]: Fixed.

**Comment by aksub99** [comment]
@hvy I have skipped the backward tests and removed unnecessary headers as suggested. The tests pass successfully now.

File: chainerx_cc/chainerx/python/routines.cc [hunk 5]
```diff
@@ -277,6 +278,17 @@ void InitChainerxCreation(pybind11::module& m) {
           "device"_a = nullptr);
 }
 
+void InitChainerxEvaluation(pybind11::module& m) {
+    // evaluation routines
+    m.def("accuracy",
```
Comment (hvy) [inline chainerx_cc/chainerx/python/routines.cc]: Could you add docs to this routine (in `chainerx/_docs/routines.py`)?

File: chainerx_cc/chainerx/python/routines.cc [hunk 6]
```diff
@@ -277,6 +278,17 @@ void InitChainerxCreation(pybind11::module& m) {
...
+          [](const ArrayBodyPtr& x1, const ArrayBodyPtr& x2, const absl::optional<int8_t>& ignore_label) {
+              return MoveArrayBody(Accuracy(Array{x1}, Array{x2}, ignore_label));
+          },
+          "x1"_a,
+          "x2"_a,
```
Comment (hvy) [inline chainerx_cc/chainerx/python/routines.cc]: Not a strong opinion but how about aligning the argument names to Chainer (`y` and `t`)? Or let me know why we shouldn't.

File: chainerx_cc/chainerx/routines/evaluation.cc [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+Array Accuracy(const Array& x1, const Array& x2, const absl::optional<int8_t>& ignore_label) {
+    if (ignore_label.has_value()) {
+        Array ignore = chainerx::FullLike(x2, Scalar{*ignore_label});
+        Array mask = Equal(x2, ignore);
+        Array ignore_cnt = Sum(mask);
+        Array pred = Where(mask, ignore, ArgMax(x1, 1).Reshape(x2.shape()));
+        Array count = Sum(Equal(pred, x2)) - ignore_cnt;
+        Scalar size{x2.GetTotalSize()};
+        Scalar total = size - AsScalar(ignore_cnt);
+        if (total == 0.0) {
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/evaluation.cc]: This should be an integer comparison?

File: chainerx_cc/chainerx/routines/evaluation.cc [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+            return Array{0}.AsType(x1.dtype());
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/evaluation.cc]: I'm not sure what constructor you're trying to call this `0` will be interpreted as a `nullptr` and will yield an error. How about using `Zeros`/`ZerosLike` ? Also, could you add tests for this code path?

File: chainerx_cc/chainerx/routines/evaluation.cc [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+#include "chainerx/routines/reduction.h"
+#include "chainerx/routines/sorting.h"
+#include "chainerx/routines/statistics.h"
+#include "chainerx/routines/type_util.h"
+#include "chainerx/scalar.h"
+#include "chainerx/shape.h"
+
+namespace chainerx {
+
+Array Accuracy(const Array& x1, const Array& x2, const absl::optional<int8_t>& ignore_label) {
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/evaluation.cc]: Let's check the dtype kind of `x1` is a float and `x2` is an integer, and throw an error otherwise.

File: chainerx_cc/chainerx/routines/evaluation.h [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+#include <cstdint>
+
+#include <absl/types/optional.h>
+
+#include "chainerx/array.h"
+
+namespace chainerx {
+
+Array Accuracy(const Array& x1, const Array& x2, const absl::optional<int8_t>& ignore_label);
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/evaluation.h]: Hm, is `int8_t` enough (max 127)? How about using `int64_t` since converting this primitive to a `Scalar` will promote it to an `int64_t` anyway and there shouldn't be any significant performance penalty?

File: tests/chainerx_tests/unit_tests/routines_tests/test_evaluation.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+                ((10, 3), (10,)),
+                ((10, 3, 1), (10,)),
+                ((10, 3, 1, 1), (10,)),
+                ((10, 3, 5), (10, 5)),
+                ((10, 3, 5, 4), (10, 5, 4)),
+                ((10, 3, 5, 4, 1), (10, 5, 4)),
+                ((10, 3, 5, 4, 1, 1), (10, 5, 4))
+            ]),
+        chainer.testing.from_pytest_parameterize(
+            'in_dtypes,out_dtype', _in_out_eval_dtypes),
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_evaluation.py]: Is the `out_dtype` read?

**Comment by hvy** [comment]
Thanks for the fix. I left some additional comments, could you take a look?

File: chainerx_cc/chainerx/python/routines.cc (hunk 5 above)
Comment (aksub99) [reply chainerx_cc/chainerx/python/routines.cc]: Fixed.

File: chainerx_cc/chainerx/python/routines.cc (hunk 6 above)
Comment (aksub99) [reply chainerx_cc/chainerx/python/routines.cc]: Fixed.

File: chainerx_cc/chainerx/routines/evaluation.cc (hunk 7 above)
Comment (aksub99) [reply chainerx_cc/chainerx/routines/evaluation.cc]: Thanks for the correction. Fixed.

File: chainerx_cc/chainerx/routines/evaluation.cc (hunk 8 above)
Comment (aksub99) [reply chainerx_cc/chainerx/routines/evaluation.cc]: I have changed it to use `Zeros` now. I have added 2 cases where this code path will be taken.
I have added a TODO to make a cleaner interface for this test though.

File: chainerx_cc/chainerx/routines/evaluation.cc (hunk 9 above)
Comment (aksub99) [reply chainerx_cc/chainerx/routines/evaluation.cc]: Sure. Modified.

File: chainerx_cc/chainerx/routines/evaluation.h (hunk 10 above)
Comment (aksub99) [reply chainerx_cc/chainerx/routines/evaluation.h]: Thanks. Made the fix.

File: tests/chainerx_tests/unit_tests/routines_tests/test_evaluation.py (hunk 11 above)
Comment (aksub99) [reply tests/chainerx_tests/unit_tests/routines_tests/test_evaluation.py]: It isn't being read. Why is this so?

**Comment by aksub99** [comment]
@hvy  Thanks for the suggestions. I have made the corrections.
I have left a question in https://github.com/chainer/chainer/pull/7526#discussion_r314941960 .
Please review.

File: tests/chainerx_tests/unit_tests/routines_tests/test_evaluation.py (hunk 11 above)
Comment (hvy) [reply tests/chainerx_tests/unit_tests/routines_tests/test_evaluation.py]: With `ChainerOpTest`, you're checking the dtype against Chainer's output dtypes, which by the way differ from the unused expected outputs that you've defined above. How did you choose the expected output dtypes in `_in_out_eval_dtypes`?

Comment (hvy) [reply tests/chainerx_tests/unit_tests/routines_tests/test_evaluation.py]: To answer your question, it's not being read because that's how `ChainerOpTest` is defined and implemented.

File: chainerx_cc/chainerx/routines/evaluation.cc (hunk 8 above)
Comment (hvy) [reply chainerx_cc/chainerx/routines/evaluation.cc]: As targets are randomly generated, you are not necessarily testing this code path just by using a fixed ignore_label. You need to construct such a case explicitly where all targets have the ignore_label.

**Comment by hvy** [comment]
Could you also add tests for empty shapes, and negative cases where you're making sure that the expected errors are raised?

File: chainerx_cc/chainerx/routines/evaluation.cc (hunk 8 above)
Comment (aksub99) [reply chainerx_cc/chainerx/routines/evaluation.cc]: @hvy I had also added shapes `(10, 1), (10, )` and `(5, 1), (5, )` for this purpose. So, when `ignore_label` is 0,  `total` will be 0 for the above 2 cases and this code path will be entered. Please let me know if I am still missing something.

Comment (aksub99) [reply chainerx_cc/chainerx/routines/evaluation.cc]: Oh yes. Sorry.
I have now set `t` to also have all zeros so that for input shapes `(10, 1)` and `(5, 1)`, this code path will be taken.

File: tests/chainerx_tests/unit_tests/routines_tests/test_evaluation.py (hunk 11 above)
Comment (aksub99) [reply tests/chainerx_tests/unit_tests/routines_tests/test_evaluation.py]: Made the fixes. Thanks for your reply.

**Comment by aksub99** [comment]
I have made the modifications. Could you please review?

File: tests/chainerx_tests/unit_tests/routines_tests/test_evaluation.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    skip_backward_test = True
+    skip_double_backward_test = True
+
+    def generate_inputs(self):
+        y, t = super().generate_inputs()
+        # TODO(aksub99): Improve tests for the case
+        # where all labels are ignored.
+        if y.shape == (10, 1) or y.shape == (5, 1):
+            self.ignore_label = 0
+            t[True] = 0
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_evaluation.py]: Is this common style for filling an array? I think the following is a bit cleaner otherwise.
```suggestion
            t.fill(self.ignore_label)
```

**Comment by aksub99** [comment]
Made the change. Thanks for the suggestions.

**Comment by aksub99** [comment]
@hvy I have added tests for invalid dtypes. I have also added a check for invalid shapes. But I am unsure how to accommodate it into the tests.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
As for invalid shapes, can you not create an additional function similar to `test_accuracy_invalid` ?

**Comment by aksub99** [comment]
I have added the required tests. Thanks for your reply. Please review:)

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Thanks for the additional tests. Could you just fix the style error reported by Jenkins https://jenkins.preferred.jp/job/chainer/job/chainer_pr/1912/TEST=CHAINERX_misc,label=mn1-p100/console ?

**Comment by aksub99** [comment]
@hvy I guess it should be fixed now.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
LGTM!

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
Jenkins, test this please.
